### PR TITLE
Support custom packager port when debugging on Android

### PR DESCRIPTION
**Pull request by ivmarkov**
This three liner addresses the issue described by Rovack here: #9935

Basically, changing the React Native Packager port from 8081 to, say, 8088 does work, *except* if you want to debug your code with ChromeDevTools (i.e. selecting "Debug JS Remotely" from the Android app dev menu).

The analysis of Rovack, pointing to a hard-coded port in DevServerHelper.getHostForJSProxy() was *spot on* - many thanks to him.

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java [hunk 1]
```diff
@@ -218,8 +218,10 @@ public String getInspectorDeviceUrl() {
   /**
    * @return the host to use when connecting to the bundle server from the host itself.
    */
-  private static String getHostForJSProxy() {
-    return AndroidInfoHelpers.DEVICE_LOCALHOST;
+  private String getHostForJSProxy() {
+    String host = getDebugServerHost();
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: What if it returns `null`?

Here's the definition:

```
public @Nullable String getDebugServerHost() {
  return mPreferences.getString(PREFS_DEBUG_SERVER_HOST_KEY, null);
}
```

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java [hunk 2]
```diff
@@ -218,8 +218,10 @@ public String getInspectorDeviceUrl() {
...
+    int portOffset = host.lastIndexOf(':');
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: Then this crashes with a `NullPointerException`.

**Comment by mkonicek** [comment]
Updated PR description - in case we end up merging this we should have a clear commit message.

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java (hunk 1 above)
Comment (ivmarkov) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: You are looking at the wrong method.

The above patch calls `DevServerHelper.getDebugServerHost()`, while you are looking at `DevInternalSettings.getDebugServerHost()`.

DevServerHelper.getDebugServerHost() cannot - ever - return null. Look at its implementation, in GitHub master:

```Java
/**
   * @return the host to use when connecting to the bundle server.
   */
  private String getDebugServerHost() {
    // Check debug server host setting first. If empty try to detect emulator type and use default
    // hostname for those
    String hostFromSettings = mSettings.getDebugServerHost();

    if (!TextUtils.isEmpty(hostFromSettings)) {
      return Assertions.assertNotNull(hostFromSettings);
    }

    String host = AndroidInfoHelpers.getServerHost();

    if (host.equals(AndroidInfoHelpers.DEVICE_LOCALHOST)) {
      FLog.w(
        ReactConstants.TAG,
        "You seem to be running on device. Run 'adb reverse tcp:8081 tcp:8081' " +
          "to forward the debug server's port to the device.");
    }

    return host;
  }
```

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java (hunk 2 above)
Comment (ivmarkov) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: See above.

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java (hunk 1 above)
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: Ah sorry. Best to check for `null` anyway in case the implementation of `getDebugServerHost` changes. Also, can `AndroidInfoHelpers.getServerHost()` return `null`?

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java (hunk 2 above)
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: Thanks!

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java [hunk 3]
```diff
@@ -218,8 +218,10 @@ public String getInspectorDeviceUrl() {
...
+    return "localhost" + (portOffset > -1? host.substring(portOffset): "");
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: nit: space after `?`

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java (hunk 1 above)
Comment (ivmarkov) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: `AndroidInfoHelpers.getServerHost()` cannot return null, as it is basically choosing amongst one of three hard-coded, non-null strings. And this method is not called directly by my "three-liner" anyway.
Also, even if I add a null-check in my method, a null host will blow up in the caller - `DevServerHelper.getJSBundleURLForRemoteDebugging()`, which relies on` getHostForJsProxy()` returning a non-null value.

TL;DR: Don't get me wrong - adding a null check is 30 seconds, but it will just not fix anything, as the whole code is written in a way where null for the host is not anticipated anywhere. But if you really insist - I'll commit a null check ASAP.

Long story: For these cases, I'm more of a fan of using proper JSR-308 `@Nullable` annotations to clearly indicate if a type includes null not. A defensive non-null check might save the current method invocation, but still blow up the call chain if the caller does not anticipate the returned value to be null - as in our case. With that said - while the ReactAndroid code does have `@Nullable` annotations here and there - it is not Java-8 compatible and does not have a JSR-308 null analyzer running as part of the build (right?). So you are also right that if `DevServerHelper.getDebugServerHost()` starts to return null for host, this change will go unnoticed at compile time.

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java (hunk 3 above)
Comment (ivmarkov) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: ... and after `:`. Will fix.

File: ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java (hunk 1 above)
Comment (ivmarkov) [inline ReactAndroid/src/main/java/com/facebook/react/devsupport/DevServerHelper.java]: Fixed the `? :` formatting and pushed a non null assertion (defensive if also possible, if you insist).

**Comment by ivmarkov** [comment]
Ping - just let me know if you still insist on a different check - I'll do it asap.

**Comment by ivmarkov** [comment]
Sorry for pushing, but this issue is seriously hindering our productivity. The change is only 3 lines of code, which had been discussed quite extensively already. I also showed willingness to rework the 3 lines (sounds ridiculous I know!) with null checking and whatnot as long as this gets merged. What else needs to be done to get this moving?

**Comment by Rovack** [comment]
@mkonicek - it's been over a month since you reviewed this PR and apparently found no significant issues with it. Is there any reason it still hasn't been merged, and that @ivmarkov hasn't received any responses in so long?

**Comment by jhen0409** [comment]
@mkonicek ping

**Comment by Timoteus78** [comment]
@mkonicek  Is there a specific reason why this hasn't been merged yet?

**Comment by valleywood** [comment]
@mkonicek Please please merge this, it's really stopning us from debugging our Android code :(

**Comment by peterheard01** [comment]
Please merge this pull request I am also having the same problems!

**Comment by peterheard01** [comment]
@Rovack did you hear back from anyone about this?

**Comment by mkonicek** [comment]
Sorry about not seeing the notifications. I don't work on React Native anymore and also there's no way to filter GitHub notifications by mentions. On GitHub any time you comment on an issue you end up getting notifications about that issue any time anyone comments. Twitter of fb groups (e.g. React Native Community) work better. You can also mention @lacker on Twitter who manages the open source team at Facebook.

screenshot 2017-05-11 14 04 11

**Comment by mkonicek** [comment]
Fixed indentation and merging.

**Comment by mkonicek** [comment]
It's a merge conflict. I'll fix it manually and will merge this. No other action is needed from you.

**Comment by ivmarkov** [comment]
@mkonicek  - thanks!

**Comment by mkonicek** [comment]
np!

**Comment by callagga** [comment]
I'm still getting this issue and just wanted to check if this fix is in place in react-native v0.49, and what I actually need to do in my test react-native project to have it point to 8080 (not 8081)?   

Background: Can't get android emulator (not device) working for a test react-native project on Windows 7. I'm doing basically:
 - react-native init AwesomeProject
 - run up emulator
 - react-native start --port 8080  [in first CMD window]
 - react-native run-android [in second CMD window]

Then on the emulator I can see its still pointing to 8081 (not 8088).

**Comment by callagga** [comment]
Discovered the answer was to on the emulator, which did have an initial view loaded do:

> Control-M (to open dev menu) > Dev Settings > Debug server host & port for device > IP:Port

[took me a while to discover the Control-M being the way to get to the menu]
